File: Data Structures & Algorithms/queue/submission-0.py
class ListNode:
    def __init__(self,val):
        self.val = val
        self.next = None
        self.prev = None

class Deque:
    
    def __init__(self):
        self.head = ListNode(-1)
        self.tail = ListNode(-1)

        self.head.next = self.tail
        self.tail.prev = self.head

    def isEmpty(self) -> bool:
        if self.head.next == self.tail:
            return True
        return False

    def append(self, value: int) -> None:
        
        new_node = ListNode(value)

        back = self.tail.prev
        back.next = new_node
        self.tail.prev = new_node
        new_node.next = self.tail
        new_node.prev = back

    def appendleft(self, value: int) -> None:
        new_node = ListNode(value)

        front = self.head.next
        front.prev = new_node
        self.head.next = new_node
        new_node.next = front
        new_node.prev = self.head

    def pop(self) -> int:
        if self.head.next == self.tail:
            return -1
        x = self.tail.prev.val
        remove = self.tail.prev
        self.tail.prev = remove.prev
        remove.prev.next = self.tail
        return x
       

    def popleft(self) -> int:

        if self.head.next == self.tail:
            return -1
        x = self.head.next.val 
        remove = self.head.next
        self.head.next = remove.next
        remove.next.prev = self.head
        return x

File: Data Structures & Algorithms/singlyLinkedList/submission-0.cpp

class Node{
    public:

    int val;
    
    Node* next{};
    Node (int val) : val(val) , next(nullptr) {}

};


class LinkedList {


private:
    Node* head;
    Node* tail;
public:
    LinkedList() : head(nullptr), tail(nullptr){}


    int get(int index) {
    
        Node* cur = head;

        for (int i =0; i< index && cur!= nullptr; i++){
            cur = cur-> next;
        }

        if (cur == nullptr) return -1;
        return cur-> val;

    }

    void insertHead(int val) {
      Node* newNode = new Node(val);
      if(head == nullptr){
        head = tail = newNode;
        return;
      }

      newNode -> next = head;
      head = newNode;

    }
    
    void insertTail(int val) {

        Node* newNode = new Node(val);
        if(tail == nullptr){
            head = tail = newNode;
            return;
        }
        tail ->next = newNode;
        tail = newNode;

    }

    bool remove(int index) {

        if(index < 0 || head == nullptr) return false;
        
        if (index == 0){
           Node* temp = head;
           head = head-> next;
           if(head == nullptr) tail = nullptr;
           delete temp;
           return true; 
        }
        
        Node* prev = head;

        for(int i =0; i < index-1; i++){
            if (prev == nullptr) return false;
            prev = prev->next;

        }

        if (prev == nullptr || prev->next == nullptr) return false;

        Node* temp = prev-> next;
        prev->next = temp->next;
        if (temp == tail) tail = prev;
        delete temp;
        return true;
    }

    vector<int> getValues() {
        vector<int> values;
        Node* cur = head;
        while (cur != nullptr){
            values.push_back(cur->val);
            cur = cur->next;
        }

        return values;
        
    }
};
